File: solutions/scala/binary-search/1/src/main/scala/BinarySearch.scala
import scala.annotation.tailrec

object BinarySearch {
  def find(list: List[Int], target: Int): Option[Int] =
    binarySearch(list, target)(0, list.length - 1)

  @tailrec
  private def binarySearch(list: List[Int], target: Int)(start: Int, end: Int): Option[Int] =
    if (start > end) {
      None
    }
    else {
      val middle = extractMiddleValue(start, end, list)
      middle.value match {
        case `target`        =>
          Some(middle.index)
        case value if value > target =>
          binarySearch(list, target)(start, middle.index - 1)
        case _               =>
          binarySearch(list, target)(middle.index + 1, end)
      }
    }

  private def extractMiddleValue(start: Int, end: Int, list: List[Int]):IndexedValue = {
    val middleIndex = start + (end - start) / 2
    val middleValue = list(middleIndex)
    IndexedValue(middleIndex, middleValue)
  }
}

File: solutions/python/binary-search/1/binary_search.py
def find(array, value):
    left_index = 0
    right_index = len(array) - 1

    while left_index <= right_index:
        middle_index = (left_index + right_index) // 2
        middle_value = array[middle_index]

        if middle_value == value:
            return middle_index

        if middle_value < value:
            left_index = middle_index + 1
        else:
            right_index = middle_index - 1

    raise ValueError("value not in array")